internal/parse/parser.go

```
@@ -113,14 +121,28 @@ func (p *Parser) add(part queryPart) {
 // a pointer to a ParsedExpr.
 func (p *Parser) Parse(input string) (*ParsedExpr, error) {
 	p.init(input)
+
 	for {
 		p.partStart = p.pos
-		if p.pos == len(p.input) {
+
+		if ip, ok, err := p.parseInputExpression(); err != nil {
+			return nil, fmt.Errorf("parser error: %s", err)
```

"cannot parse expression: " would be a better prefix here, as it avoids the idea that the parser itself is in error. But I suggest adding it at the top level with something like this as it can handle it generally:
```
func (p *Parser) Parse(input string) (expr *ParsedExpr, err error) {
    defer func() {
        if err != nil {
            err = fmt.Errorf("cannot parse expression: %w", err)
        }
    }()
```
This way the rest of the logic can ignore the issue altogether.

---

This is a nice approach. Though would this not mean we ignore errors until we have consumed the whole input string? It would wait until the Parse function has finished before dealing with errors so if there was a problem right at the start of the input we would still try and parse the rest rather than giving up when the error occurred.

Parsing should be relatively low cost compared to the other things we are doing though so this could be fine.

---

I've changed the errors to use wrapping and the %w. I've held off adding the defer function - waiting on your reply to the previous comment